feat: Unify Base `AuthClient` Options

danielbankhead opened the pull request:
Also, expose `Transporter` options

Fixes #1621 🦕

d-goog committed:
feat: Expose `Gaxios` instance and default

d-goog committed:
feat: Unify Base `AuthClient` Options

d-goog committed:
docs: clean up documentation

d-goog committed:
chore: Discourage external use via `@internal`

d-goog committed:
refactor: minor

d-goog committed:
refactor: clean up `transporter` interface, options, and docs

d-goog committed:
Merge branch 'main' into authclient-enhancements

d-goog committed:
Merge branch 'main' of github.com:googleapis/google-auth-library-nodejs into authclient-enhancements

d-goog committed:
Merge branch 'main' of github.com:googleapis/google-auth-library-nodejs into authclient-enhancements

d-goog committed:
Merge branch 'main' of github.com:googleapis/google-auth-library-nodejs into authclient-enhancements

d-goog committed:
test: Init tests for `AuthClient`

d-goog committed:
fix: Use entire `JWT` to create accurate `createScoped`

d-goog committed:
chore: update `typescript` in fixtures

d-goog committed:
test: Use Sinon Fake Timer to Avoid Flaky Time Issue (#1667)

* test: Use Sinon Fake Timer to Avoid Flaky Time Issue

* chore: change order for clarity

d-goog committed:
Merge branch 'main' into authclient-enhancements

aeitzman commented on src/auth/baseexternalclient.ts:
```diff
@@ -169,9 +160,10 @@ export abstract class BaseExternalAccountClient extends AuthClient {
   constructor(
     options: BaseExternalAccountClientOptions,
-    additionalOptions?: RefreshOptions
+    additionalOptions?: AuthClientOptions
```
This seems very confusing to me now that the BaseExternalAccountClientOptions object is a child of AuthClientOptions, doesn't seem like it will be clear to users what should be where. It would make sense to me to create a new constructor that takes in just 1 options objects (BaseExternalAccountClientOptions) and does the existing constructor logic, and then keep this constructor for backwards compatibility but just have it pass the merged options objects to the new constructor like you are calling the super here. And then have comments explaining that the old constructor is deprecated/only for compatability

danielbankhead replied:
I wish I was around for the naming 😅 `BaseExternalAccountClient` is the base of external, non-Google Cloud accounts (AWS, Azure, etc.). Since all `AuthClient`s are required to have `transporter`, `projectId`, `quotaProjectId`, and now `universeDomain` properties it would make since for this to be handled in the root class.

aeitzman replied:
Just realized we can't have multiple constructors in typescript so what I was suggesting doesn't make sense anyway. I agree that it make sense to handle it all in the base class. My concern is that having both options objects will be confusing to users since they are both AuthClientOptions. At the very least can you add to the constructor documentation that the additional options param is deprecated and the options can all be passed via the BaseExternalAccountClientOptions parameter? Other option is if we just wait to release this until the breaking change suggested by  #1672 then we won't be in this weird in between state at all.

danielbankhead replied:
Docs-deprecated `additionalOptions`, as the [@deprecated](https://tsdoc.org/pages/tags/deprecated/) tag is considered breaking as it would produce build warnings.

d-goog committed:
Merge branch 'main' into authclient-enhancements

d-goog committed:
docs(sample): Improve `keepAlive` sample with `transporterOptions`

aeitzman commented on src/auth/authclient.ts:
```diff
@@ -13,31 +13,91 @@
+ * Base Auth Client configuration
+ */
+export type AuthClientOptions = Partial<
+  OriginalAndCamel<AuthJSONOptions> & {
```
So this options object will end up having each of the AuthJSONOptions twice? Once with key in camelCase and one with the key in snake_case?

danielbankhead replied:
Yep, with type hints and documentation. This way we can pass in the raw JSON properties from a file as-is or provide the options in a conventional JS pattern.

aeitzman commented on src/auth/authclient.ts:
```diff
@@ -88,16 +148,40 @@ export abstract class AuthClient
+    super();
+
+    // Shared auth options
+    this.projectId = getOpt(opts, 'project_id') ?? null;
```
Why alias the getOriginalOrCamel here?  Made it more confusing for me to figure out what was actually going on here.

danielbankhead replied:
That's fair, unaliased it

aeitzman commented on src/auth/idtokenclient.ts:
```diff
@@ -41,7 +46,7 @@ export class IdTokenClient extends OAuth2Client {
    */
   constructor(options: IdTokenOptions) {
-    super();
+    super(options);
```
This is new functionality right? Looks like the IdTokenClient didn't previously support RefreshClient options. Are we sure it should be used here?

danielbankhead replied:
I thought it would be cleaner to pass the parameters directly to it's parent; however we mostly care about setting `AuthClientOptions`. I figured it wouldn't hurt to expose underlying `OAuth2ClientOptions`, as it wouldn't harm `IdTokenClient`'s functionality. Thoughts?

Also, I noticed the `OAuth2ClientOptions` interface was accidentally duplicated from a merge; removed the duplicated interface to avoid confusion.

aeitzman commented on test/test.util.ts:
```diff
@@ -50,18 +61,23 @@ describe('util', () => {
 
+      // just set, so should be fine
       lru.set('third', 3);
+      assert.equal(lru.get('third'), 3);
```
Is this intended to be part of this PR?

danielbankhead replied:
Accident; somehow the target wasn't `main`: #1667

aeitzman commented on samples/keepalive.js:
```diff
@@ -30,21 +30,26 @@ const https = require('https');
  * Acquire a client, and make a request to an API that's enabled by default.
  */
 async function main() {
+  // create a new agent with keepAlive enabled
```
nit: Add period on the comment

d-goog committed:
docs: Docs-deprecate `additionalOptions`

danielbankhead commented on src/auth/authclient.ts:
```diff
@@ -88,16 +148,40 @@ export abstract class AuthClient
+    this.universeDomain = getOpt(opts, 'universe_domain') ?? DEFAULT_UNIVERSE;
+
+    // Shared client options
+    this.transporter = opts.transporter ?? new DefaultTransporter();
```
A `Transporter` is an interface, which `gaxios` satisfies by having a `request` method with a matching signature. I think this makes it a lot easier for customers that simply want to pass the default `gaxios` or custom `Gaxios` instance w/o building/maintaining a transporter.

d-goog committed:
refactor: un-alias `getOriginalOrCamel`

d-goog committed:
chore: remove confusing duplicate interface

d-goog committed:
docs: nit

d-goog committed:
Merge branch 'main' into authclient-enhancements

d-goog committed:
Merge branch 'main' of github.com:googleapis/google-auth-library-nodejs into authclient-enhancements

d-goog committed:
docs: Improve camelCased option documentation

We can refactor once https://github.com/microsoft/TypeScript/issues/50715 has been resolved.

d-goog committed:
refactor: Unify `OriginalAndCamel` & `SnakeToCamelObject`

+ make recursive

d-goog committed:
docs: Add issue tracking link

d-goog committed:
refactor: Replace `getOriginalOrCamel` with a cleaner API

d-goog committed:
feat: Allow optional `obj`

d-goog committed:
refactor: re-add `SnakeToCamelObject`

d-goog committed:
refactor: dedupe interface for now

d-goog committed:
feat: Add camelCase options for `IdentityPoolClient`
